// @unovis/vue@1.3.5/html-components/bullet-legend/index.js
import { defineComponent as i, computed as s, ref as r, onMounted as d, nextTick as m, onUnmounted as f, watch as v, openBlock as _, createElementBlock as b } from "vue";
import { BulletLegend as x } from "@unovis/ts";
import { useForwardProps as k, arePropsEqual as g } from "../../utils/props.js";
const w = /* @__PURE__ */ i({
  __name: "index",
  props: {
    items: {},
    labelClassName: {},
    onLegendItemClick: { type: Function },
    labelFontSize: {},
    labelMaxWidth: {},
    bulletSize: {},
    bulletShape: { type: [Function, null] },
    data: {}
  },
  setup(c, { expose: p }) {
    const l = c;
    s(() => l.data);
    const n = k(l), t = r(), o = r();
    return d(() => {
      m(() => {
        o.value && (t.value = new x(o.value, n.value));
      });
    }), f(() => {
      var e;
      (e = t.value) == null || e.destroy();
    }), v(n, (e, a) => {
      var u;
      g(e, a) || (u = t.value) == null || u.update(n.value);
    }), p({
      component: t
    }), (e, a) => (_(), b("div", {
      "data-vis-bullet-legend": "",
      ref_key: "elRef",
      ref: o
    }, null, 512));
  }
});
export {
  w as default
};
