// @unovis/vue@1.3.5/components/line/index.js
import { defineComponent as s, inject as d, computed as m, ref as y, onMounted as v, nextTick as f, onUnmounted as F, watch as l, openBlock as h, createElementBlock as x } from "vue";
import { Line as _ } from "@unovis/ts";
import { useForwardProps as b, arePropsEqual as g } from "../../utils/props.js";
import { componentAccessorKey as k } from "../../utils/context.js";
const A = { "data-vis-component": "" }, w = /* @__PURE__ */ s({
  __name: "index",
  props: {
    color: { type: [Function, String, Array, null] },
    curveType: {},
    lineWidth: {},
    lineDashArray: { type: [Function, null] },
    fallbackValue: {},
    highlightOnHover: { type: Boolean },
    cursor: { type: [Function, String, null] },
    x: { type: [Function, Number, null] },
    y: { type: [Function, Number, null, Array] },
    id: { type: Function },
    xScale: { type: [Object, Function] },
    yScale: { type: [Object, Function] },
    excludeFromDomainCalculation: { type: Boolean },
    duration: {},
    events: {},
    attributes: {},
    data: {}
  },
  setup(i, { expose: p }) {
    const c = i, n = d(k), o = m(() => n.data.value ?? c.data), a = b(c), t = y();
    return v(() => {
      f(() => {
        var e;
        t.value = new _(a.value), (e = t.value) == null || e.setData(o.value), n.update(t.value);
      });
    }), F(() => {
      var e;
      (e = t.value) == null || e.destroy(), n.destroy();
    }), l(a, (e, r) => {
      var u;
      g(e, r) || (u = t.value) == null || u.setConfig(a.value);
    }), l(o, () => {
      var e;
      (e = t.value) == null || e.setData(o.value);
    }), p({
      component: t
    }), (e, r) => (h(), x("div", A));
  }
});
export {
  w as default
};
