// @unovis/vue@1.3.5/components/stacked-bar/index.js
import { defineComponent as s, inject as d, computed as m, ref as y, onMounted as v, nextTick as b, onUnmounted as f, watch as i, openBlock as x, createElementBlock as F } from "vue";
import { StackedBar as _ } from "@unovis/ts";
import { useForwardProps as g, arePropsEqual as h } from "../../utils/props.js";
import { componentAccessorKey as B } from "../../utils/context.js";
const S = { "data-vis-component": "" }, j = /* @__PURE__ */ s({
  __name: "index",
  props: {
    color: { type: [Function, String, Array, null] },
    barWidth: {},
    barMaxWidth: {},
    dataStep: {},
    barPadding: {},
    roundedCorners: { type: [Number, Boolean] },
    cursor: { type: [Function, String, null] },
    barMinHeight1Px: { type: Boolean },
    barMinHeightZeroValue: {},
    orientation: {},
    x: { type: [Function, Number, null] },
    y: { type: [Function, Number, null, Array] },
    id: { type: Function },
    xScale: { type: [Object, Function] },
    yScale: { type: [Object, Function] },
    excludeFromDomainCalculation: { type: Boolean },
    duration: {},
    events: {},
    attributes: {},
    data: {}
  },
  setup(l, { expose: p }) {
    const r = l, o = d(B), n = m(() => o.data.value ?? r.data), a = g(r), t = y();
    return v(() => {
      b(() => {
        var e;
        t.value = new _(a.value), (e = t.value) == null || e.setData(n.value), o.update(t.value);
      });
    }), f(() => {
      var e;
      (e = t.value) == null || e.destroy(), o.destroy();
    }), i(a, (e, c) => {
      var u;
      h(e, c) || (u = t.value) == null || u.setConfig(a.value);
    }), i(n, () => {
      var e;
      (e = t.value) == null || e.setData(n.value);
    }), p({
      component: t
    }), (e, c) => (x(), F("div", S));
  }
});
export {
  j as default
};
