// @unovis/vue@1.3.5/components/axis/index.js
import { defineComponent as s, inject as d, computed as y, ref as m, onMounted as x, nextTick as k, onUnmounted as v, watch as r, openBlock as F, createElementBlock as T } from "vue";
import { Axis as f } from "@unovis/ts";
import { useForwardProps as B, arePropsEqual as b } from "../../utils/props.js";
import { axisAccessorKey as _ } from "../../utils/context.js";
const g = { "data-vis-axis": "" }, M = /* @__PURE__ */ s({
  __name: "index",
  props: {
    position: {},
    type: {},
    fullSize: { type: Boolean },
    label: {},
    labelFontSize: {},
    labelMargin: {},
    labelColor: {},
    gridLine: { type: Boolean },
    tickLine: { type: Boolean },
    domainLine: { type: Boolean },
    minMaxTicksOnly: { type: Boolean },
    tickFormat: { type: Function },
    tickValues: {},
    numTicks: {},
    tickTextFitMode: {},
    tickTextWidth: {},
    tickTextSeparator: {},
    tickTextForceWordBreak: { type: Boolean },
    tickTextTrimType: {},
    tickTextFontSize: {},
    tickTextAlign: {},
    tickTextColor: {},
    tickPadding: {},
    x: { type: [Function, Number, null] },
    y: { type: [Function, Number, null, Array] },
    id: { type: Function },
    color: { type: [Function, String, Array, null] },
    xScale: { type: [Object, Function] },
    yScale: { type: [Object, Function] },
    excludeFromDomainCalculation: { type: Boolean },
    duration: {},
    events: {},
    attributes: {},
    data: {}
  },
  setup(u, { expose: p }) {
    const o = u, n = d(_), a = y(() => n.data.value ?? o.data), i = B(o), t = m();
    return x(() => {
      k(() => {
        var e;
        t.value = new f(i.value), (e = t.value) == null || e.setData(a.value), n.update(t.value);
      });
    }), v(() => {
      var e;
      (e = t.value) == null || e.destroy(), n.destroy(o.type);
    }), r(i, (e, c) => {
      var l;
      b(e, c) || (l = t.value) == null || l.setConfig(i.value);
    }), r(a, () => {
      var e;
      (e = t.value) == null || e.setData(a.value);
    }), p({
      component: t
    }), (e, c) => (F(), T("div", g));
  }
});
export {
  M as default
};
