// @unovis/vue@1.3.5/components/sankey/index.js
import { defineComponent as d, inject as s, computed as y, ref as b, onMounted as g, nextTick as m, onUnmounted as F, watch as u, openBlock as h, createElementBlock as S } from "vue";
import { Sankey as v } from "@unovis/ts";
import { useForwardProps as f, arePropsEqual as x } from "../../utils/props.js";
import { componentAccessorKey as C } from "../../utils/context.js";
const k = { "data-vis-component": "" }, L = /* @__PURE__ */ d({
  __name: "index",
  props: {
    id: { type: Function },
    heightNormalizationCoeff: {},
    exitTransitionType: {},
    enterTransitionType: {},
    highlightSubtreeOnHover: { type: Boolean },
    highlightDuration: {},
    highlightDelay: {},
    iterations: {},
    nodeSort: { type: [Function, null] },
    linkSort: { type: [Function, null] },
    nodeWidth: {},
    nodeAlign: {},
    nodeHorizontalSpacing: {},
    nodeMinHeight: {},
    nodeMaxHeight: {},
    nodePadding: {},
    showSingleNode: { type: Boolean },
    nodeCursor: { type: [Function, String, null] },
    nodeIcon: { type: [Function, String, null] },
    nodeColor: { type: [Function, String, Array, null] },
    nodeFixedValue: { type: [Function, Number, null] },
    nodeIconColor: { type: [Function, String, Array, null] },
    linkColor: { type: [Function, String, null] },
    linkValue: { type: [Function, Number, null] },
    linkCursor: { type: [Function, String, null] },
    label: { type: [Function, String, null] },
    subLabel: { type: [Function, String, null] },
    labelPosition: { type: [Function, null] },
    labelVerticalAlign: {},
    labelBackground: { type: Boolean },
    labelFit: {},
    labelMaxWidth: {},
    labelExpandTrimmedOnHover: { type: Boolean },
    labelTrimMode: {},
    labelFontSize: {},
    labelTextSeparator: {},
    labelForceWordBreak: { type: Boolean },
    labelColor: { type: [Function, String, Array, null] },
    labelCursor: { type: [Function, String, null] },
    labelVisibility: { type: Function },
    subLabelFontSize: {},
    subLabelColor: { type: [Function, String, Array, null] },
    subLabelPlacement: {},
    subLabelToLabelInlineWidthRatio: {},
    duration: {},
    events: {},
    attributes: {},
    data: {}
  },
  setup(c, { expose: p }) {
    const i = c, t = s(C), o = y(() => t.data.value ?? i.data), l = f(i), n = b();
    return g(() => {
      m(() => {
        var e;
        n.value = new v(l.value), (e = n.value) == null || e.setData(o.value), t.update(n.value);
      });
    }), F(() => {
      var e;
      (e = n.value) == null || e.destroy(), t.destroy();
    }), u(l, (e, a) => {
      var r;
      x(e, a) || (r = n.value) == null || r.setConfig(l.value);
    }), u(o, () => {
      var e;
      (e = n.value) == null || e.setData(o.value);
    }), p({
      component: n
    }), (e, a) => (h(), S("div", k));
  }
});
export {
  L as default
};
