// jsonc-eslint-parser/lib/parser/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseForESLint = void 0;
const espree_1 = require("./modules/espree");
const visitor_keys_1 = require("./visitor-keys");
const convert_1 = require("./convert");
const token_store_1 = require("./token-store");
const semver_1 = require("semver");
const extend_parser_1 = require("./extend-parser");
const DEFAULT_ECMA_VERSION = "latest";
function parseForESLint(code, options) {
    var _a, _b, _c, _d;
    const parserOptions = Object.assign({ filePath: "<input>", ecmaVersion: DEFAULT_ECMA_VERSION }, options || {}, {
        loc: true,
        range: true,
        raw: true,
        tokens: true,
        comment: true,
        eslintVisitorKeys: true,
        eslintScopeManager: true,
    });
    parserOptions.ecmaVersion = normalizeEcmaVersion(parserOptions.ecmaVersion);
    const ctx = getJSONSyntaxContext(parserOptions.jsonSyntax);
    const tokens = [];
    const comments = [];
    const tokenStore = new token_store_1.TokenStore(tokens);
    const n
