// @unovis/vue@1.3.5/components/grouped-bar/index.js
import { defineComponent as s, inject as d, computed as m, ref as y, onMounted as v, nextTick as f, onUnmounted as g, watch as i, openBlock as x, createElementBlock as F } from "vue";
import { GroupedBar as b } from "@unovis/ts";
import { useForwardProps as _, arePropsEqual as h } from "../../utils/props.js";
import { componentAccessorKey as B } from "../../utils/context.js";
const S = { "data-vis-component": "" }, w = /* @__PURE__ */ s({
  __name: "index",
  props: {
    color: { type: [Function, String, Array, null] },
    groupWidth: {},
    groupMaxWidth: {},
    dataStep: {},
    groupPadding: {},
    barPadding: {},
    roundedCorners: { type: [Number, Boolean] },
    barMinHeight: {},
    cursor: { type: [Function, String, null] },
    orientation: {},
    x: { type: [Function, Number, null] },
    y: { type: [Function, Number, null, Array] },
    id: { type: Function },
    xScale: { type: [Object, Function] },
    yScale: { type: [Object, Function] },
    excludeFromDomainCalculation: { type: Boolean },
    duration: {},
    events: {},
    attributes: {},
    data: {}
  },
  setup(p, { expose: l }) {
    const r = p, o = d(B), n = m(() => o.data.value ?? r.data), a = _(r), t = y();
    return v(() => {
      f(() => {
        var e;
        t.value = new b(a.value), (e = t.value) == null || e.setData(n.value), o.update(t.value);
      });
    }), g(() => {
      var e;
      (e = t.value) == null || e.destroy(), o.destroy();
    }), i(a, (e, u) => {
      var c;
      h(e, u) || (c = t.value) == null || c.setConfig(a.value);
    }), i(n, () => {
      var e;
      (e = t.value) == null || e.setData(n.value);
    }), l({
      component: t
    }), (e, u) => (x(), F("div", S));
  }
});
export {
  w as default
};
